File: benchmarks/expr_eval.go
package main

import "fmt"

// Go has no sum types, so an AST is a tagged struct with pointer children and a
// kind switch — the idiomatic equivalent of witchy's `type Expr` + `match`.
type Expr struct {
	kind byte // 0 = Num, 1 = Add, 2 = Mul
	n    int
	a, b *Expr
}

func num(n int) *Expr      { return &Expr{kind: 0, n: n} }
func add(a, b *Expr) *Expr { return &Expr{kind: 1, a: a, b: b} }
func mul(a, b *Expr) *Expr { return &Expr{kind: 2, a: a, b: b} }

func eval(e *Expr) int {
	switch e.kind {
	case 0:
		return e.n
	case 1:
		return eval(e.a) + eval(e.b)
	default:
		return eval(e.a) * eval(e.b)
	}
}

func build(depth int) *Expr {
	if depth <= 0 {
		return num(1)
	}
	return add(mul(build(depth-1), num(2)), build(depth-1))
}

func main() {
	total := 0
	for i := 0; i < 10; i++ {
		total += eval(build(16))
	}
	fmt.Println(total)
}

File: benchmarks/expr_eval.rb
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

class Num
  attr_reader :n
  def initialize(n); @n = n; end
  def eval; @n; end
end

class Add
  def initialize(a, b); @a = a; @b = b; end
  def eval; @a.eval + @b.eval; end
end

class Mul
  def initialize(a, b); @a = a; @b = b; end
  def eval; @a.eval * @b.eval; end
end

def build(depth)
  if depth <= 0
    Num.new(1)
  else
    Add.new(Mul.new(build(depth - 1), Num.new(2)), build(depth - 1))
  end
end

total = 0
10.times do
  total += build(16).eval
end
puts total
puts "bench_ns=#{Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - t0}"
